packages/compass-aggregations/src/modules/aggregation.ts

```diff
@@ -39,32 +101,120 @@ export const runAggregation = (): ThunkAction<
+> => {
+  return (dispatch) => dispatch(fetchAggregationData(1));
+};
+
+export const fetchPrevPage = (): ThunkAction<
+  void,
+  RootState,
+  void,
+  Actions
+> => {
+  return (dispatch, getState) => {
+    const {
+      aggregation: { page }
+    } = getState();
+    if (page <= 1) {
+      return;
+    }
+    dispatch(fetchAggregationData(page - 1));
+  };
+};
+
+export const fetchNextPage = (): ThunkAction<
+  void,
+  RootState,
+  void,
+  Actions
+> => {
+  return (dispatch, getState) => {
+    const {
+      aggregation: { isLast, page }
+    } = getState();
+    if (isLast) {
+      return;
+    }
+    dispatch(fetchAggregationData(page + 1));
+  };
+};
+
+export const cancelAggregation = (): ThunkAction<
+  void,
+  RootState,
+  void,
+  Actions
+> => {
+  return (_dispatch, getState) => {
+    const {
+      aggregation: { abortController },
+    } = getState();
+    abortController?.abort();
+  };
+};
+
+const fetchAggregationData = (page: number): ThunkAction<
+  void,
+  RootState,
+  void,
+  Actions
 > => {
   return async (dispatch, getState) => {
     const {
       pipeline,
       namespace,
       maxTimeMS,
       collation,
-      dataService: { dataService }
+      dataService: { dataService },
+      aggregation: { limit },
     } = getState();
 
-    const stages = pipeline.map(generateStage);
-    const options: AggregateOptions = {
-      maxTimeMS: maxTimeMS || DEFAULT_MAX_TIME_MS,
-      allowDiskUse: true,
-      collation: collation || undefined,
-    };
-    if (dataService) {
+    if (!dataService) {
+      return;
+    }
+
+    const abortController = new AbortController();
+    const signal = abortController.signal;
+    abortController.signal.addEventListener('abort', cancelAggregation, { once: true });
+
+    try {
+      dispatch({
+        type: ActionTypes.AggregationStarted,
+        abortController,
+      });
+
+      const stages = pipeline.map(generateStage);
+      const options: AggregateOptions = {
+        maxTimeMS: maxTimeMS || DEFAULT_MAX_TIME_MS,
+        allowDiskUse: true,
+        collation: collation || undefined,
+      };
       const cursor = dataService.aggregate(
         namespace,
         stages,
         options
-      );
-      const documents = await cursor.toArray();
-      return dispatch({ type: ActionTypes.RunAggregation, documents });
+      ).skip((page - 1) * limit).limit(limit);
+
+      if (!await raceWithAbort(cursor.hasNext(), signal)) {
+        return dispatch({ type: ActionTypes.LastPageReached });
+      }
+
+      const documents = await raceWithAbort(cursor.toArray(), signal);
+      dispatch({
+        type: ActionTypes.AggregationFinished,
+        documents,
+        page,
+        isLast: documents.length < limit,
+      });
+    } catch (e) {
+      dispatch({
+        type: ActionTypes.AggregationFailed,
+        error: (e as Error).message,
+      });
+    } finally {
+      abortController.signal.removeEventListener('abort', cancelAggregation);
```

Comment:
Should we also implement the session-based logic for stopping the aggregations server-side that we use in compass-crud? Otherwise a user cancellation will stop the operation in the UI but not on the server afaict.